### PR TITLE
[Feature] Support new parameter - EBNF in xgrammar

### PR DESCRIPTION
**Pull request by adarshxs**
## Motivation
XGrammar enables efficient structured generation. XGrammar backend already support enforcement of EBNF (Extended Backus-Naur Form) grammars.
adds feature requested in: https://github.com/sgl-project/sglang/issues/2376

## Modifications
- added a new parameter called `ebnf` in sampling_params.py and treat it similar to regex and JSON.
- Implemented the `test_ebnf_generate` method to validate multiple valid output patterns based on a complex EBNF grammar.

## Checklist

- [x] Format your code according to the [Contributor Guide](https://github.com/sgl-project/sglang/blob/main/docs/references/contributor_guide.md).
- [x] Add unit tests as outlined in the [Contributor Guide](https://github.com/sgl-project/sglang/blob/main/docs/references/contributor_guide.md).
- [ ] Update documentation as needed, including docstrings or example tutorials. 

TODO: update docs

### CODE REVIEW

**Comment by zhyncs** [comment]
cc @Ubospica @DarkSharpness

File: test/srt/test_ebnf_constrained.py [hunk 1]
```diff
@@ -0,0 +1,247 @@
...
+
+        self.run_decode(
+            ebnf=self.__class__.ebnf_grammar,
+            expected_patterns=allowed_patterns,
+            prompt=prompt,
+            n=3,
+        )
+
+
+class TestJumpForward(unittest.TestCase):
```
Comment (merrymercy) [inline test/srt/test_ebnf_constrained.py]: Should this be `class TestJumpForward(TestEBNFConstrained):` ?
The current one does not include any test cases.

**Review by merrymercy — COMMENTED** [review]
Great work! Can you also support it with the OpenAI API? https://github.com/sgl-project/sglang/blob/19ba2b0ea9241bf6a3ff5918916cd725755d1b6b/test/srt/test_openai_server.py#L470-L480

File: test/srt/test_ebnf_constrained.py (hunk 1 above)
Comment (adarshxs) [reply test/srt/test_ebnf_constrained.py]: yes my bad it should be `class TestJumpForward(TestEBNFConstrained):` 
fixing it

**Comment by adarshxs** [comment]
@merrymercy done

File: python/sglang/srt/managers/scheduler.py [hunk 2]
```diff
@@ -589,12 +589,15 @@ def handle_generate_request(
...
             or req.sampling_params.regex is not None
+            or req.sampling_params.ebnf is not None
         ):
             assert self.grammar_backend is not None
             if req.sampling_params.json_schema is not None:
                 key = ("json", req.sampling_params.json_schema)
             elif req.sampling_params.regex is not None:
                 key = ("regex", req.sampling_params.regex)
+            elif req.sampling_params.ebnf is not None:
+                key = ("ebnf", req.sampling_params.ebnf)
```
Comment (Ubospica) [inline python/sglang/srt/managers/scheduler.py:601]: Shall we raise a warning when multiple format is specified at the same time?

**Review by Ubospica — COMMENTED** [review]
It overall looks great for me. Thanks!

File: python/sglang/srt/managers/scheduler.py (hunk 2 above)
Comment (adarshxs) [reply python/sglang/srt/managers/scheduler.py:601]: yes we raise an error here for that:
https://github.com/sgl-project/sglang/blob/fda6936b28f8805b02db6064b785b7480697ae32/python/sglang/srt/sampling/sampling_params.py#L116-L122

works like this:
image

**Comment by merrymercy** [comment]
Thanks for the contribution! It is merged @adarshxs

**Comment by adarshxs** [comment]
Thank you!

**Pull request merged**